File: packages/shared-types/src/spatial.ts
import { z } from "zod";

// --- 3D Point ---
export const Point3DSchema = z.object({
  x: z.number(),
  y: z.number(),
  z: z.number(),
});
export type Point3D = z.infer<typeof Point3DSchema>;

// --- 4x4 Homogeneous Transformation Matrix (row-major) ---
export const TransformMatrixSchema = z.array(z.number()).length(16);
export type TransformMatrix = z.infer<typeof TransformMatrixSchema>;

// --- Raycast Request ---
export const RaycastRequestSchema = z.object({
  device_id: z.string(),
  direction: Point3DSchema,
  bim_model_id: z.string(),
  max_distance: z.number().positive().default(100.0),
});
export type RaycastRequest = z.infer<typeof RaycastRequestSchema>;

// --- Spatial Hit (single ray-BIM intersection) ---
export const SpatialHitSchema = z.object({
  asset_id: z.string(),
  asset_name: z.string(),
  ifc_type: z.string(),
  hit_point: Point3DSchema,
  distance: z.number(),
  confidence: z.number().min(0).max(1),
});
export type SpatialHit = z.infer<typeof SpatialHitSchema>;

// --- Raycast Response ---
export const RaycastResponseSchema = z.object({
  ray_origin: Point3DSchema,
  ray_direction: Point3DSchema,
  hits: z.array(SpatialHitSchema),
});
export type RaycastResponse = z.infer<typeof RaycastResponseSchema>;

// --- Triangulation Request ---
export const TriangulationRequestSchema = z.object({
  observations: z.array(
    z.object({
      device_id: z.string(),
      direction: Point3DSchema,
      confidence: z.number().min(0).max(1),
    })
  ).min(2),
  bim_model_id: z.string(),
});
export type TriangulationRequest = z.infer<typeof TriangulationRequestSchema>;

// --- Triangulation Result ---
export const TriangulationResultSchema = z.object({
  estimated_point: Point3DSchema,
  residual_error: z.number(),
  nearest_asset: SpatialHitSchema.nullable(),
  contributing_devices: z.array(z.string()),
});
export type TriangulationResult = z.infer<typeof TriangulationResultSchema>;

// --- Spatial Query (batch: N predictions -> ranked asset hits) ---
export const SpatialQueryRequestSchema = z.object({
  predictions: z.array(
    z.object({
      device_id: z.string(),
      class_name: z.string(),
      confidence: z.number(),
      vector: z.tuple([z.number(), z.number(), z.number()]),
    })
  ),
  bim_model_id: z.string(),
  top_k: z.number().int().positive().default(5),
});
export type SpatialQueryRequest = z.infer<typeof SpatialQueryRequestSchema>;

// --- Calibration Matrix ---
export const CalibrationMatrixSchema = z.object({
  device_id: z.string(),
  bim_model_id: z.string(),
  matrix: TransformMatrixSchema,
  updated_at: z.string().datetime(),
});
export type CalibrationMatrix = z.infer<typeof CalibrationMatrixSchema>;

File: cloud/api-gateway/src/api_gateway/routers/spatial.py
"""Spatial router — proxy to spatial engine for DOA→asset hit queries."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

router = APIRouter()


class SpatialQueryRequest(BaseModel):
    device_id: str
    direction: dict  # {x, y, z}
    bim_model_id: str
    max_distance: float = 100.0


class TriangulationRequest(BaseModel):
    observations: list[dict]  # [{device_id, direction: {x,y,z}, confidence}]
    bim_model_id: str


@router.post("/spatial/query")
async def spatial_query(request: SpatialQueryRequest):
    """Query spatial engine: DOA vector → ranked BIM asset hits."""
    # TODO: Forward to spatial engine gRPC service
    return {
        "ray_origin": {"x": 0, "y": 0, "z": 0},
        "ray_direction": request.direction,
        "hits": [],
        "message": "Spatial engine not connected yet",
    }


@router.post("/spatial/triangulate")
async def triangulate(request: TriangulationRequest):
    """Multi-sensor triangulation: multiple DOA observations → 3D point."""
    if len(request.observations) < 2:
        raise HTTPException(status_code=400, detail="At least 2 observations required")

    # TODO: Forward to spatial engine gRPC service
    return {
        "estimated_point": {"x": 0, "y": 0, "z": 0},
        "residual_error": 0.0,
        "nearest_asset": None,
        "contributing_devices": [o["device_id"] for o in request.observations],
    }
